src/librustc_mir/dataflow/generic/mod.rs

```diff
@@ -0,0 +1,305 @@
+/// functions for each statement in this way, the transfer function for an entire basic block can
+/// be computed efficiently.
+///
+/// `Analysis` is automatically implemented for all implementers of `GenKillAnalysis`.
+pub trait GenKillAnalysis<'tcx>: Analysis<'tcx> {
+    /// See `Analysis::apply_statement_effect`.
+    fn statement_effect(
+        &self,
+        trans: &mut impl GenKill<Self::Idx>,
+        statement: &mir::Statement<'tcx>,
+        location: Location,
+    );
+
+    /// See `Analysis::apply_before_statement_effect`.
+    fn before_statement_effect(
+        &self,
+        _trans: &mut impl GenKill<Self::Idx>,
+        _statement: &mir::Statement<'tcx>,
+        _location: Location,
+    ) {}
+
+    /// See `Analysis::apply_terminator_effect`.
+    fn terminator_effect(
+        &self,
+        trans: &mut impl GenKill<Self::Idx>,
+        terminator: &mir::Terminator<'tcx>,
+        location: Location,
+    );
+
+    /// See `Analysis::apply_before_terminator_effect`.
+    fn before_terminator_effect(
+        &self,
+        _trans: &mut impl GenKill<Self::Idx>,
+        _terminator: &mir::Terminator<'tcx>,
+        _location: Location,
+    ) {}
+
+    /// See `Analysis::apply_call_return_effect`.
+    fn call_return_effect(
+        &self,
+        trans: &mut impl GenKill<Self::Idx>,
+        block: BasicBlock,
+        func: &mir::Operand<'tcx>,
+        args: &[mir::Operand<'tcx>],
+        return_place: &mir::Place<'tcx>,
+    );
+}
+
+impl<A> Analysis<'tcx> for A
+where
+    A: GenKillAnalysis<'tcx>
+{
+    fn apply_statement_effect(
+        &self,
+        state: &mut BitSet<Self::Idx>,
+        statement: &mir::Statement<'tcx>,
+        location: Location,
+    ) {
+        self.statement_effect(state, statement, location);
+    }
+
+    fn apply_before_statement_effect(
+        &self,
+        state: &mut BitSet<Self::Idx>,
+        statement: &mir::Statement<'tcx>,
+        location: Location,
+    ) {
+        self.before_statement_effect(state, statement, location);
+    }
+
+    fn apply_terminator_effect(
+        &self,
+        state: &mut BitSet<Self::Idx>,
+        terminator: &mir::Terminator<'tcx>,
+        location: Location,
+    ) {
+        self.terminator_effect(state, terminator, location);
+    }
+
+    fn apply_before_terminator_effect(
+        &self,
+        state: &mut BitSet<Self::Idx>,
+        terminator: &mir::Terminator<'tcx>,
+        location: Location,
+    ) {
+        self.before_terminator_effect(state, terminator, location);
+    }
+
+    fn apply_call_return_effect(
+        &self,
+        state: &mut BitSet<Self::Idx>,
+        block: BasicBlock,
+        func: &mir::Operand<'tcx>,
+        args: &[mir::Operand<'tcx>],
+        return_place: &mir::Place<'tcx>,
+    ) {
+        self.call_return_effect(state, block, func, args, return_place);
+    }
+}
+
+/// The legal operations for a transfer function in a gen/kill problem.
+pub trait GenKill<T>: Sized {
```

Comment:
I'm only on the first commit, so I don't really know the answer to this question yet: Does abstracting `GenKill<T>` into a trait make sense? Its hard to imagine a significantly different alternative implementation to what the existing `GenKillSet` struct does, apart from the details of how the two bitsets are represented, I suppose.

Reply:
Okay I reached the end of this commit and now see that you have implemented `GenKill` for `BitSet`. I do not think this is a good idea.

Specifically, you have not stated in the documentation for `trait GenKill` that the underlying gen or kill sets may be *missing*, depending on how the trait is implemented. Having `fn kill()` just remove the bit from the gen-set does not really capture the semantics of a dataflow analysis: the clients parameterized over `GK: GenKill` are going to have to say "You must use `GenKillSet` for this to work", or they are going to have to ensure that their particular analysis does not actually rely ever generating a kill effect (in which case, I would have `<BitSet as GenKill>::kill` panic instead of solely clearing the bit in the gen-set.

Reply:
(I suppose another way it could hypothetically make sense to `impl GenKill for BitSet` would be for an analysis where the kill-set is *always* the set-complement of the gen-set. But AFAICT that is not what you implemented, and I am not aware of any analysis where the gen/kill sets satisfy that property.)

Reply:
In `ResultsCursor`, the `apply` methods are applied directly to the `BitSet` containing the dataflow state at the cursor position. The `GenKill` trait is meant to abstract over two different contexts where we need to apply the transfer function. When coalescing transfer functions for an entire block before running dataflow analysis, we are storing a function (`trans`) that can be applied later, which requires two bitsets. However, when seeking with a dataflow cursor, we don't care about the function itself; we only need to know its effect on the state vector, (`trans(state)`). There's no need to store extra bitsets.

There are other ways to avoid allocating here, but having a proper trait is the cleanest approach IMO. I agree that the documentation should be clearer about what is happening when in the `GenKill` implementation for `BitSet`.

Reply:
Okay. I'm sorry that I did not read this response more carefully back when you first posted it, as it would have saved me the effort of rediscovering the motivation for this abstraction myself. (But then again, it did force you to embed the motivation into the doc-comment for the trait, so maybe my oversight was a win in the long run...)